### ISSUE TITLE
v4.0.x/v4.1.x: Configure stage produces truncated include/library paths

### ISSUE THREAD

**Issue by eschnett** [open-mpi/ompi#9162]
Thank you for taking the time to submit an issue!

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

4.1.1 and 4.0.6

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

I install OpenMPI via Spack. This downloads a release tarball.

### If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.

n/a

### Please describe the system on which you are running

* Operating system/version: macOS 11.4
* Computer hardware: Apple
* Network type: Wireless (should not be relevant)

-----------------------------

## Details of the problem

Please describe, in detail, the problem that you are having, including the behavior you expect to see, the actual behavior that you are seeing, steps to reproduce the problem, etc.  It is most helpful if you can attach a small program that a developer can use to reproduce your problem.

OpenMPI produces truncated paths in its generated pkgconfig files, for example:
```
# Copyright (c) 2010 Cisco Systems, Inc.  All rights reserved.
# Copyright (c) 2016      IBM Corporation.  All rights reserved.
#
Name: Open MPI
Description: Portable high-performance MPI implementation
Version: 4.1.1
URL: http://www.open-mpi.org/
#
prefix=/Users/eschnett/src/spack/opt/spack/darwin-bigsur-skylake/gcc-11.1.0/openmpi-4.1.1-4rhxaejgvtzajqrvrs3ikl2z5lg7zyiw
exec_prefix=${prefix}
includedir=${prefix}/include
libdir=${exec_prefix}/lib
pkgincludedir=${includedir}/openmpi
#
# Note that -lopen-pal and all the EXTRA_LIBS are only necessary when
# static linking (they're pulled in by libopen-rte.so's implicit
# dependencies), so only list these in Libs.private.
#
Libs: -L${libdir} -L/Users/eschnett/src/spack/opt/spack/darwin-bigsur-skylake/gcc-11.1.0/hwloc-2.5.0-fdhuxgmzrrihnqfemkfgsnfvis6opoam/lib -L/Users/eschnett/src/spack/opt/spack/darwin-bigsur-skylake/gcc-11.1.0/libevent-2.1.12-kmrptsjbl4ghvwa54yzyozt7xqusvjac/lib -L/Users/eschnett/src/spack/opt/spack/darwin-bigsur-skylake/gcc-11.1.0/zlib-1.2.11-oll4m2dh7d3vjpdlfpncdmeihx562kjr/lib -L/Users/eschnett/src/spa  -Wl,-rpath,/Users/eschnett/src/spack/opt/spack/darwin-bigsur-skylake/apple-clang-12.0.0/gcc-11.1.0-mwjrgc7ul5pr72tpkf7t6yt4x624ujzw/lib/gcc/x86_64-apple-darwin20.5.0/11.1.0 -Wl,-rpath,/Users/eschnett/src/spack/opt/spack/darwin-bigsur-skylake/apple-clang-12.0.0/gcc-11.1.0-mwjrgc7ul5pr72tpkf7t6yt4x624ujzw/lib -Wl,-rpath,/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/lib  -lmpi
Libs.private: -lopen-rte -lopen-pal -lhwloc -levent_core -levent_pthreads -lz 
#
Cflags: -I${includedir}      
```

Notice the path `-L/Users/eschnett/src/spa` in the `Libs` setting. This is an incomplete path; most of the path has been cut off.

This problem is already visible in the output of `configure`. Here are the first lines that I could find where a path is truncated (lines 2716 ff in my output):
```
PMIx configuration:
-----------------------
Version: 3.2.3
Debug build: no
Platform file: (none)

External Packages
-----------------------
HWLOC: yes (embedded)
Libevent: yes (embedded)
ZLIB: yes (/Users/eschnett/src/spack/opt/spack/darwin-bigsur-skylake/gcc-11.1.0/zlib-1.2.11-oll4m2dh7d3vjpdlfpncdmeihx562kjr)
 
configure: /bin/sh '/var/folders/gb/x5lhfpj15ln66g6br549tz5m0000gn/T/eschnett/spack-stage/spack-stage-openmpi-4.1.1-4rhxaejgvtzajqrvrs3ikl2z5lg7zyiw/spack-src/opal/mca/pmix/pmix3x/pmix/configure' succeeded for opal/mca/pmix/pmix3x/pmix
checking if v3.x component is to be used... yes - using the internal v3.x library
checking PMIx extra wrapper CPPFLAGS...   -I/Users/eschnett/src/spack/opt/spack/darwin-bigsur-skylake/gcc-11.1.0/zlib-1.2.11-oll4m2dh7d3vjpdlfpncdmeihx562kjr  -I/Users/eschnett/src/spack/o
checking PMIx extra wrapper LDFLAGS...   -L/Users/eschnett/src/spack/opt/spack/darwin-bigsur-skylake/gcc-11.1.0/zlib-1.2.11-oll4m2dh7d3vjpdlfpncdmeihx562kjr/lib  -L/Users/eschnett/src/spa
checking PMIx extra wrapper LIBS...  -lz -lz
checking if MCA component pmix:pmix3x can compile... yes
configure: WARNING: ignoring pmix_pmix3x_WRAPPER_EXTRA_CPPFLAGS (  -I/Users/eschnett/src/spack/opt/spack/darwin-bigsur-skylake/gcc-11.1.0/zlib-1.2.11-oll4m2dh7d3vjpdlfpncdmeihx562kjr  -I/Users/eschnett/src/spack/o): component conditions not met
```

Notice the settings `-I/Users/eschnett/src/spack/o` and `-L/Users/eschnett/src/spa`, which are both truncated.

Here are some details about how Spack configures OpenMPI:
[spack-build-01-autoreconf-out.txt](https://github.com/open-mpi/ompi/files/6849819/spack-build-01-autoreconf-out.txt)
[spack-build-env.txt](https://github.com/open-mpi/ompi/files/6849824/spack-build-env.txt)
[spack-configure-args.txt](https://github.com/open-mpi/ompi/files/6849823/spack-configure-args.txt)
[spack-build-02-configure-out.txt](https://github.com/open-mpi/ompi/files/6849821/spack-build-02-configure-out.txt)

**Comment by eschnett** [comment]
I just notice that the configure output above says
```
HWLOC: yes (embedded)
Libevent: yes (embedded)
```

This looks wrong; both are provided as external libraries (see the configure arguments).

**Comment by jsquyres** [comment]
@eschnett I filed #9170 about the hwloc/libevent external vs. embedded issue -- even though that fell out of your initial report, it's a separate issue, so let's follow up on that separately.

**Comment by ggouaillardet** [comment]
> I just notice that the configure output above says
> 
> ```
> HWLOC: yes (embedded)
> Libevent: yes (embedded)
> ```

Note these lines belong to PMIx

Open MPI lines are at the end and correctly report external packages

**Comment by jsquyres** [comment]
> Note these lines belong to PMIx

@ggouaillardet I opened #9170 about that issue.  I'm trying to replicate the truncated paths issue and will post more back here when I have additional info.

**Comment by eschnett** [comment]
I case it's relevant: I have Xcode 12.4 installed on my system. There is also Xcode 12.5 (I had to downgrade because something wasn't working right) and Xcode 12.5.1.

**Comment by jsquyres** [comment]
I am able to reproduce this with an Open MPI v4.1.1 tarball in very specific conditions:

1. Here's the minimal `configure` command line that I found that will cause the problem:
   ```
   ./configure --enable-static --with-zlib=$base/zlib-1.2.11/install
   ```
2. The zlib directory must be "long".  Specifically, in the above `configure` command line:
    * This triggers the problem: `base=/Users/jsquyres/tmp/this-is-a-really-long-dir/lets-make-it-extra-long/heres-another-subdir/how-long-can-we-go`
    * This does not trigger the problem: `base=/Users/jsquyres/tmp/zlib-1.2.11/install`

I've run out of time today to investigate further, but the above suggests that `--with-zlib` processing is where the problem originates (perhaps in https://github.com/open-mpi/ompi/blob/v4.1.x/config/opal_setup_zlib.m4?).

**Comment by rhc54** [comment]
Could also be in PMIx - we use it there too.

**Comment by jsquyres** [comment]
Aren't all issues categorized as PMIx problems until they can be proven otherwise?  :stuck_out_tongue:

**Comment by rhc54** [comment]
Yeah sure - blame the retired guy! 🏖️

**Comment by jsquyres** [comment]
v4.1.x fix in #9176.  v4.0.x will come after merging to v4.1.x.

**Comment by jsquyres** [comment]
Fixed on both branches.  Closing this issue.  Thanks!
